### PR TITLE
(#12) Page Not Found for Disease

### LINKED ISSUE #12: Story: "404" Screen for when the disease does not exist.
## As a user, I want to know that I entered the wrong disease code by being presented with a 404 page so that I can check my URL and ensure that I entered the correct information.

Use the 404 page that we use in other apps (glossary, drug dictionary, etc.)

> **ESTIMATE** 2

### Acceptance Criteria

```gherkin
Scenario: Page Not Found displays when a user views a dynamic listing page with invalid disease parameter
Given user navigates to “/C123455”
And "trialListingPageType" is set to "Disease"
When the user has entered an incorrect disease
Then the system redirects users to a 404 page 
     Then page title on error page is "Page Not Found"
     And the search bar appears below the text
     And the following links and texts exist on the page: "We can't find the page you're looking for." Visit the homepage, browse by cancer type, or use the search below." "Have a question? Get in touch."
  
      | https://www.cancer.gov          | homepage      | 
      | https://www.cancer.gov/types    | cancer type   | 
      | https://www.cancer.gov/contact  | Get in touch  |
And the system returns a 404 status code
```
```gherkin
Given the application returns a "Page not found" message
 Then the title tag should be "Page Not Found"
  Then the page should contain the markup <meta name="prerender-status-code" content="404">
```
```gherkin
Scenario: Page Load Analytics fires for a 404 with invalid disease
 Given "trialListingPageType" is set to "Disease"
    And "baseHost" is set to "http://localhost:3000"
    And "canonicalHost" is set to "https://www.cancer.gov"
    And "analyticsPublishedDate" is set to "02/02/2011"
    And "siteName" is set to "National Cancer Institute"
 When the user navigates to "/c123456"
   Then page title on error page is "Page Not Found"
    And browser waits
   Then there should be an analytics event with the following details

| key                                          | value                                            |
| type                                         | PageLoad                                         |
| event                                        | TrialListingPageApp:Load:PageNotFound            |
| page.name                                    | www.cancer.gov/c123456                           |
| page.title                                   | Page Not Found                                   |
| page.metaTitle                               | Page Not Found                                   |
| page.language                                | english                                          |
| page.type                                    | nciAppModulePage                                 |
| page.channel                                 | Clinical Trials                                  |
| page.contentGroup                            | Clinical Trials: Custom                          |
| page.additionalDetails.trialListingPageType  | disease                                          |
```
### Resources:
* Feature: JIRA Ticket #
* Mockups: *Here goes a URL to or the name of the mockup(s) in inVision*
* Web Page: *Here goes a URL to an existing web page*

### Notes
Adding markup will allow a pre-renderer to detect the error and report the correct 404 status code.

*Some complementary notes if necessary:*

* > Here goes a quote from an email
* Here goes whatever useful information can exist…

---

## Solution
As had been designed, the higher order component that fetches the listing info will be responsible for showing the 404 when one of the parameters that is fetched returns a 404 from the API.

### Prerequisites
- *required dependency for identifying the problem, scope and then solution for this issue*
- [ ] Short Spike Description - Issue #9999

### Tasks
- [ ] Add logic to `CTLViewsHoC` to display `PageNotFound` for 404s
- [ ] Update unit test for `CTLViewsHoC`
- [ ] Add integration tests 

### Technical Notes
- More notes, or technical nuances/direction

### PR DESCRIPTION
**Pull request by kate-mashkina**
- enabled PageNotFound for 404 status
- added prerender meta tag
- added integration tests

Closes #12

### CODE REVIEW

**Comment by kate-mashkina** [comment]
@VictoriaSunNIH @mworrest  @alinai here are links for review : https://react-app-dev.cancer.gov/clinical-trials-listing-app/pr-40/c456667?cfg=0 and https://react-app-dev.cancer.gov/clinical-trials-listing-app/pr-40/chicken-nugget?cfg=0

**Comment by kate-mashkina** [comment]
@mworrest as we discussed, search bar and submit button are now full width at <641 px - please re-check the PR https://react-app-dev.cancer.gov/clinical-trials-listing-app/pr-40/c456667?cfg=0

**Comment by mworrest** [comment]
Looks good--thanks!

**Comment by alinai** [comment]
Looks good. Thanks @kate-mashkina!

**Review by bryanpizzillo — CHANGES_REQUESTED** [review]
Please fix the proxy to support known 404 urls that should respond with a 404, and not litter the test output with file not found errors.

The example from the output:
```
[Error: ENOENT: no such file or directory, open '/home/runner/work/clinical-trials-listing-app/clinical-trials-listing-app/support/mock-data/listing-information/c123455.json'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/home/runner/work/clinical-trials-listing-app/clinical-trials-listing-app/support/mock-data/listing-information/c123455.json'
}
No listing-information mock found for request.
    ✓ Page Load Analytics fires for a 404 on a disease dynamic listing page with non-existent c-code (2292ms)
[Error: ENOENT: no such file or directory, open '/home/runner/work/clinical-trials-listing-app/clinical-trials-listing-app/support/mock-data/listing-information/chicken-nugget.json'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/home/runner/work/clinical-trials-listing-app/clinical-trials-listing-app/support/mock-data/listing-information/chicken-nugget.json'
}
```

Additionally I see another error in the logs:
(node:3270) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 preprocessor:close listeners added to [EventEmitter]. Use emitter.setMaxListeners() to increase limit

**Comment by kate-mashkina** [comment]
@bryanpizzillo I took care of the proxy. 
The other error mentioned appears to be an obvious Cypress's issue and still outstanding (see here https://github.com/cypress-io/cypress/issues/1305) - but the good thing is that Cypress assures it does not cause things to crash.
I've done more digging and the same issue was also present in drug dictionary tests( precisely at the 10th feature it would be logged... so it just happens that pageNotFound is the 10th)

**Review by seyilonge-nci — APPROVED** [review]

**Review by sarinapadilla — APPROVED** [review]

File: support/mock-data/listing-information/c123455-404.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{"Message":"Could not find codes 'c12345'."}
```
Comment (arcepaul) [inline support/mock-data/listing-information/c123455-404.json:1]: blecch.  So this came up in the app scan as having needed to be changed.  @sarinapadilla those changes haven't been made yet, right?

**Review by bryanpizzillo — APPROVED** [review]

**Review by arcepaul — APPROVED** [review]

**Pull request merged**